cityofaustin/atd-moped
# Persist toggled columns when filtering project list view

## Associated issues

https://github.com/cityofaustin/atd-data-tech/issues/8931

Two things are happening in this PR that involved changing parts of Material Table.

One is persisting columns. Columns are toggled in the material table toolbar (https://github.com/material-table-core/core/blob/master/src/components/MTableToolbar/index.js#L193) by selecting a checkbox. The state of the columns visible or hidden are handled in the Material Table's data manager, its internal state handler. I tried changing the `onColumnsChanged` function, but found that I really needed to do was work with the onChange function in the checkbox.  I took the MTableToolBar source code and pared it down to only the columns selection. 

Further enhancement is to store the column configuration either in a cookie, local storage or in the url. 

While I was testing the columns persisting, I noticed a bug when searching/filtering that you can replication on staging right now. MT handles its page rendering differently when using remote data vs local data, [remote data docs](https://material-table.com/#/docs/features/remote-data). (We have it configured as local data.) If you search or filter and get for example 4 results, when you clear the filter, you will only see 4 records instead of all the records. 

I am more than happy to walk through this in person, but the gist of what the bug was is this: 
https://github.com/material-table-core/core/blob/master/src/utils/data-manager.js#L1238
The data manager is rendering whatever is returned from this function. If pagination is true, then it will slice the array and only show page size. So even though we are getting new data when we reset the filter, it's only showing the previous length's amount of data. 

Setting pagination as false will prevent our pagination components from showing up, so that's a no go. Another option is to change the data configuration to be remote data configuration. But that resulted in the table not updating, I tracked it down to https://github.com/material-table-core/core/blob/master/src/material-table.js#L200, where the data manager wouldn't compare the prevProps to this.props and then as a result not update the renderState. 

The solution I found was to override the renderData prop in MTableBody, to circumvent the data manager's sliced array. 

## Testing
**URL to test:** 

test on moped test or run locally

**Steps to test:**

To test the persisting columns, toggle some columns off and then search / filter for something. The columns you toggled before the search should not change. However, navigating away and returning will reset the columns to default. 

To test the search bug -- search or filter something with few results. clear the search and see that everything comes back. 



---
#### Ship list
